File: packages/calorie-cortisol-tool/shared/python/cc_contracts/contracts.py
"""REST webhook contracts — Python mirror of `contracts/webhooks.ts`.

Shared payload/response shapes for the lab-results and FHIR webhooks. HMAC
verification, structural validation, and persistence live in later tasks.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Literal, Optional

LabResultPayloadFormat = Literal["HL7", "JSON"]
LabResultsIngestStatus = Literal["accepted", "results-pending", "rejected"]
FhirVersion = Literal["R4"]
FhirIngestStatus = Literal["accepted", "rejected"]


@dataclass
class LabResultReading:
    """A single normalized cortisol result carried in a lab webhook (Req 8.4/8.5)."""

    sample_id: str
    collected_at: str
    value: float
    unit: str
    time_of_day_bucket: Optional[str] = None  # diurnal bucket (Req 8.3)


@dataclass
class LabResultsWebhookRequest:
    """`POST /webhooks/lab-results` request body (Req 8.4)."""

    order_id: str
    lab_partner_id: str
    format: LabResultPayloadFormat
    raw_message: Optional[str] = None
    readings: Optional[List[LabResultReading]] = None


@dataclass
class LabResultsWebhookResponse:
    """`POST /webhooks/lab-results` response envelope (Req 8.4/8.8)."""

    status: LabResultsIngestStatus
    order_id: str
    accepted_count: int
    rejected_count: int
    reason: Optional[str] = None


@dataclass
class FhirWebhookRequest:
    """`POST /webhooks/fhir` request body — Epic MyChart FHIR R4 bundle (Req 14.6)."""

    version: FhirVersion
    order_id: str
    bundle: Dict[str, object] = field(default_factory=dict)


@dataclass
class FhirWebhookResponse:
    """`POST /webhooks/fhir` response envelope (Req 14.6/14.7)."""

    status: FhirIngestStatus
    order_id: str
    imported_count: int
    reason: Optional[str] = None

File: packages/calorie-cortisol-tool/shared/go/contracts.go
package contracts

// REST webhook contracts — Go mirror of contracts/webhooks.ts. Shared
// payload/response shapes for the lab-results and FHIR webhooks. HMAC
// verification, structural validation, and persistence live in later tasks.

// LabResultPayloadFormat is a payload encoding accepted by the lab-results
// webhook (Req 8.4).
type LabResultPayloadFormat string

const (
	LabFormatHL7  LabResultPayloadFormat = "HL7"
	LabFormatJSON LabResultPayloadFormat = "JSON"
)

// LabResultsIngestStatus is an outcome state of a lab-results webhook (Req 8.4/8.8).
type LabResultsIngestStatus string

const (
	LabResultsAccepted LabResultsIngestStatus = "accepted"
	LabResultsPending  LabResultsIngestStatus = "results-pending"
	LabResultsRejected LabResultsIngestStatus = "rejected"
)

// FhirVersion is a FHIR resource version accepted by the FHIR webhook (Req 14.6).
type FhirVersion string

// FhirR4 is the supported FHIR version.
const FhirR4 FhirVersion = "R4"

// FhirIngestStatus is an outcome state of a FHIR import (Req 14.6/14.7).
type FhirIngestStatus string

const (
	FhirAccepted FhirIngestStatus = "accepted"
	FhirRejected FhirIngestStatus = "rejected"
)

// LabResultReading is a single normalized cortisol result carried in a lab
// webhook (Req 8.4/8.5).
type LabResultReading struct {
	SampleID        string          `json:"sampleId"`
	CollectedAt     string          `json:"collectedAt"`
	Value           float64         `json:"value"`
	Unit            string          `json:"unit"`
	TimeOfDayBucket TimeOfDayBucket `json:"timeOfDayBucket,omitempty"` // diurnal bucket (Req 8.3)
}

// LabResultsWebhookRequest is the POST /webhooks/lab-results request body (Req 8.4).
type LabResultsWebhookRequest struct {
	OrderID      string                 `json:"orderId"`
	LabPartnerID string                 `json:"labPartnerId"`
	Format       LabResultPayloadFormat `json:"format"`
	RawMessage   string                 `json:"rawMessage,omitempty"`
	Readings     []LabResultReading     `json:"readings,omitempty"`
}

// LabResultsWebhookResponse is the POST /webhooks/lab-results response envelope
// (Req 8.4/8.8).
type LabResultsWebhookResponse struct {
	Status        LabResultsIngestStatus `json:"status"`
	OrderID       string                 `json:"orderId"`
	AcceptedCount int                    `json:"acceptedCount"`
	RejectedCount int                    `json:"rejectedCount"`
	Reason        string                 `json:"reason,omitempty"`
}

// FhirWebhookRequest is the POST /webhooks/fhir request body — an Epic MyChart
// SMART-on-FHIR R4 bundle plus the linking order (Req 14.6).
type FhirWebhookRequest struct {
	Version FhirVersion            `json:"version"`
	OrderID string                 `json:"orderId"`
	Bundle  map[string]interface{} `json:"bundle"`
}

// FhirWebhookResponse is the POST /webhooks/fhir response envelope (Req 14.6/14.7).
type FhirWebhookResponse struct {
	Status        FhirIngestStatus `json:"status"`
	OrderID       string           `json:"orderId"`
	ImportedCount int              `json:"importedCount"`
	Reason        string           `json:"reason,omitempty"`
}
